### PR TITLE
Seperate ad test

### PR DESCRIPTION
**Pull request by zhouyx**

### CODE REVIEW

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    testApiHandler = null;
+  });
+
+  it('embed-state API', () => {
+    expect(testApiHandler.embedStateApi_).length.to.be.null;
+    testIframe.setAttribute(
+        'data-amp-3p-sentinel', 'amp3ptest');
+    testApiHandler.startUp(testIframe, true);
+    insert(testIframe);
+    expect(testApiHandler.embedStateApi_).to.not.be.null;
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: please avoid accessing private member in test

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    testAd = null;
+    testApiHandler = null;
+  });
+
+  it('embed-state API', () => {
+    expect(testApiHandler.embedStateApi_).length.to.be.null;
+    testIframe.setAttribute(
+        'data-amp-3p-sentinel', 'amp3ptest');
+    testApiHandler.startUp(testIframe, true);
+    insert(testIframe);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: Can we move those lines to `beforeEach`. We are probably going to need them in every test case.

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import {createIframePromise} from '../../../../testing/iframe';
+import * as sinon from 'sinon';
+
+describe('amp-ad-api-handler', () => {
+  let sandbox;
+  let container;
+  let testIframe;
+  let testAdElement;
+  let testAd;
+  let testApiHandler;
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: we can remove the `test` prefix of those variable names.

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  const iframeSrc = 'http://ads.localhost:' + location.port +
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: let's remove the protocol (`http:`) in the URL.

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            '/base/test/fixtures/served/iframe.html';
+  function insert(iframe) {
+    container.doc.body.appendChild(iframe);
+  }
+
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    testAdElement = document.createElement('amp-ad');
+    testAd = new AmpAd3PImpl(testAdElement);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: let's not depend on `AmpAd3PImpl`, use a mock instead.

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    testApiHandler = new AmpAdApiHandler(testAd, testAd.element);
+    return createIframePromise().then(c => {
+      container = c;
+      const i = c.doc.createElement('iframe');
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: seems this intermediate variable is not necessary, you can do `testIframe = c.doc.createElement('iframe');`

**Comment by zhouyx** [comment]
@lannka fixed PTAL

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import {BaseElement} from '../../../../src/base-element';
+import {createIframePromise} from '../../../../testing/iframe';
+import * as sinon from 'sinon';
+
+describe('amp-ad-api-handler', () => {
+  let sandbox;
+  let container;
+  let iframe;
+  let adElement;
+  let ad;
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: suggested naming: `adImpl`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    sandbox.restore();
+    adElement = null;
+    ad = null;
+    apiHandler = null;
+  });
+
+  it('embed-state API', () => {
+    return new Promise(resolve => {
+      apiHandler.sendEmbedInfo_ = () => {
+        resolve(apiHandler);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: is it necessary to call resolve with the `apiHandler`?

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      };
+      iframe.onload = function() {
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: use `() => {}`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        iframe.contentWindow.postMessage({
+          sentinel: 'amp-test',
+          type: 'subscribeToEmbedState',
+          is3p: true,
+          amp3pSentinel: 'amp3ptest',
+        }, '*');
+        apiHandler.iframe_.src = iframeSrc;
+      };
+      expect(apiHandler.embedStateApi_.clientWindows_.length).to.equal(0);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: expect(clientWindows).to.be.empty

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    }).then(apiHandler => {
+      expect(apiHandler.embedStateApi_.clientWindows_.length).to.equal(1);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: `expect(clientWindows).to.be.have.length(1)`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  let apiHandler;
+  const iframeSrc = '//ads.localhost:' + location.port +
+            '/base/test/fixtures/served/iframe.html';
+  function insert(iframe) {
+    container.doc.body.appendChild(iframe);
+  }
+
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    adElement = document.createElement('amp-ad');
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: seems not used, can be a local variable?

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    ad = new BaseElement(adElement);
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: ditto

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js (hunk 11 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: Can't do this. I received error like 
image

Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: It worked for length = 1 though. But when I tried `to.have.length(2)` I get error message like this. guess has something to do w/ `clientWindows_`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js (hunk 8 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: Not really

**Comment by zhouyx** [comment]
fixed PTAL

**Comment by lannka** [comment]
LGTM

**Pull request merged**